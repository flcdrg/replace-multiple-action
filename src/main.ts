import * as core from '@actions/core';
import { promises as fs } from 'fs';
import { replaceInstances } from './replaceInstances';
import { glob } from 'glob';

export type IReplacements = {
  find: string;
  replace: string;
}[];

async function run(): Promise<void> {
  try {
    core.info('starting');
    const files: string = core.getInput('files', { required: true });
    const find: string = core.getInput('find', { required: true });

    const encoding: BufferEncoding = core.getInput(
      'encoding'
    ) as BufferEncoding;

    const prefix: string = core.getInput('prefix');
    const suffix: string = core.getInput('suffix');

    const findData: IReplacements = JSON.parse(find);

    // options is optional
    const matchedFiles = await glob(files, {});

    for (const file of matchedFiles) {
      const originalContent = await fs.readFile(file, encoding);
      let content = originalContent;

      content = replaceInstances(findData, content, prefix, suffix);

      if (originalContent !== content) {
        await fs.writeFile(file, content, encoding);
        core.info(`Updating ${file}`);
      }
    }
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message);
  }
}

run();
