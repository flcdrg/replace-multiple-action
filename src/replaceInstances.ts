import { IReplacements } from './main';

export function replaceInstances(
  findData: IReplacements,
  content: string,
  prefix: string,
  suffix: string
): string {
  for (const pair of findData) {
    const pattern =
      prefix +
      // escape the find pattern
      // eslint-disable-next-line no-useless-escape
      pair.find.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&') +
      suffix;

    // Handle when prefix and/or suffix are empty
    if (prefix !== '' && suffix !== '') {
      content = content.replace(
        new RegExp(pattern, 'g'),
        `$1${pair.replace}$2`
      );
    }
    if (prefix === '' && suffix !== '') {
      content = content.replace(new RegExp(pattern, 'g'), `${pair.replace}$1`);
    }
    if (prefix !== '' && suffix === '') {
      content = content.replace(new RegExp(pattern, 'g'), `$1${pair.replace}`);
    }
    if (prefix === '' && suffix === '') {
      content = content.replace(new RegExp(pattern, 'g'), pair.replace);
    }
  }
  return content;
}
