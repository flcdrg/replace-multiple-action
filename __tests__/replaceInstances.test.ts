import { expect, test } from '@jest/globals';
import { IReplacements } from '../src/main';
import { replaceInstances } from '../src/replaceInstances';

const prefix = '(^|\\s+|\\()';
const suffix = '($|\\s+|\\))';

test('space separated', () => {
  const data: IReplacements = [
    {
      find: 'http://localhost',
      replace: 'https://wayback.org/http://localhost'
    }
  ];

  const content = 'some text http://localhost blah blah';

  const result = replaceInstances(data, content, prefix, suffix);

  expect(result).toMatchSnapshot();
});

test('start and end', () => {
  const data: IReplacements = [
    {
      find: 'http://localhost',
      replace: 'https://wayback.org/http://localhost'
    }
  ];

  const content = 'http://localhost blah \nblah and http://localhost';

  const result = replaceInstances(data, content, prefix, suffix);

  expect(result).toMatchSnapshot();
});

test('in a link', () => {
  const data: IReplacements = [
    {
      find: 'http://localhost',
      replace: 'https://wayback.org/http://localhost'
    }
  ];

  const content = 'some [text](http://localhost) blah blah';

  const result = replaceInstances(data, content, prefix, suffix);

  expect(result).toMatchSnapshot();
});

test('not links', () => {
  const data: IReplacements = [
    {
      find: 'http://localhost',
      replace: 'https://wayback.org/http://localhost'
    }
  ];

  const content =
    'some [text]http://localhost) blah blah and this -http://localhost';

  const result = replaceInstances(data, content, prefix, suffix);

  expect(result).toMatchSnapshot();
});

test('empty prefix and suffix', () => {
  const data: IReplacements = [
    {
      find: 'find',
      replace: 'replaceemptyprefixsuffix'
    }
  ];

  const content = 'some text find blah blah';

  const result = replaceInstances(data, content, '', '');

  expect(result).toMatchSnapshot();
});

test('empty prefix', () => {
  const data: IReplacements = [
    {
      find: 'find',
      replace: 'replaceemtpyprefix'
    }
  ];

  const content = 'some text find blah blah';

  const result = replaceInstances(data, content, '(text )', '');

  expect(result).toMatchSnapshot();
});

test('empty suffix', () => {
  const data: IReplacements = [
    {
      find: 'find',
      replace: 'replaceemptysuffix'
    }
  ];

  const content = 'some text find blah blah';

  const result = replaceInstances(data, content, '', '( blah)');

  expect(result).toMatchSnapshot();
});

test('multiple replacements', () => {
  const prefix = '';
  const suffix = '';

  const data: IReplacements = [
    {
      find: 'DB_USER=',
      replace: 'DB_USER=FozzieBear',
    },
    {
      find: 'DB_PASSWORD=',
      replace: 'DB_PASSWORD=WockaWocka',
    }
  ];

  const content = 'DB_HOST=\nDB_USER=\nDB_PASSWORD=\n';

  const result = replaceInstances(data, content, prefix, suffix);

  expect(result).toMatchSnapshot();
});