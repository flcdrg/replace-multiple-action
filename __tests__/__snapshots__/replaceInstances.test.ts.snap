// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty prefix 1`] = `"some text replaceemtpyprefix blah blah"`;

exports[`empty prefix and suffix 1`] = `"some text replaceemptyprefixsuffix blah blah"`;

exports[`empty suffix 1`] = `"some text replaceemptysuffix blah blah"`;

exports[`in a link 1`] = `"some [text](https://wayback.org/http://localhost) blah blah"`;

exports[`not links 1`] = `"some [text]http://localhost) blah blah and this -http://localhost"`;

exports[`space separated 1`] = `"some text https://wayback.org/http://localhost blah blah"`;

exports[`start and end 1`] = `
"https://wayback.org/http://localhost blah 
blah and https://wayback.org/http://localhost"
`;
